homeassistant/components/evohome/entity.py

```diff
@@ -111,101 +110,77 @@ def current_temperature(self) -> float | None:
-            )
+        key = "temp" if isinstance(self._evo_device, evo.Zone) else "state"
+
+        self._setpoints = {
+            "this_sp_from": this_sp_dtm,
+            f"this_sp_{key}": this_sp_val,
+            "next_sp_from": next_sp_dtm,
+            f"next_sp_{key}": next_sp_val,
+        }
 
         return self._setpoints
 
-    async def _update_schedule(self) -> None:
+    async def _update_schedule(self, force_refresh: bool = False) -> None:
         """Get the latest schedule, if any."""
 
-        assert isinstance(self._evo_device, evo.HotWater | evo.Zone)  # mypy check
-
-        try:
-            schedule = await self._evo_broker.call_client_api(
-                self._evo_device.get_schedule(), update_state=False
-            )
-        except evo.InvalidSchedule as err:
-            _LOGGER.warning(
-                "%s: Unable to retrieve a valid schedule: %s",
-                self._evo_device,
-                err,
+        async def get_schedule() -> None:
+            try:
+                schedule = await self.coordinator.call_client_api(
+                    self._evo_device.get_schedule(),  # type: ignore[arg-type]
+                    request_refresh=False,
+                )
+            except evo.InvalidScheduleError as err:
+                _LOGGER.warning(
+                    "%s: Unable to retrieve a valid schedule: %s",
+                    self._evo_device,
+                    err,
+                )
+                self._schedule = {}
+                return
+            else:
+                self._schedule = schedule or {}  # mypy hint
+
+            _LOGGER.debug("Schedule['%s'] = %s", self.name, schedule)
+
+        if (
+            force_refresh
+            or self._schedule is None
+            or (
+                (until := self._setpoints.get("next_sp_from")) is not None
+                and until < datetime.now(UTC)
             )
-            self._schedule = {}
-        else:
-            self._schedule = schedule or {}
+        ):  # must use self._setpoints, not self.setpoints
+            await get_schedule()
+
+        _ = self.setpoints  # update the setpoints attr
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        super()._handle_coordinator_update()
 
-        _LOGGER.debug("Schedule['%s'] = %s", self.name, self._schedule)
+        self._device_state_attrs[SZ_ACTIVE_FAULTS] = self._evo_device.active_faults
+        self._device_state_attrs["setpoints"] = self._setpoints
 
-    async def async_update(self) -> None:
-        """Get the latest state data."""
-        next_sp_from = self._setpoints.get("next_sp_from", "2000-01-01T00:00:00+00:00")
-        next_sp_from_dt = dt_util.parse_datetime(next_sp_from)
-        if next_sp_from_dt is None or dt_util.now() >= next_sp_from_dt:
-            await self._update_schedule()  # no schedule, or it's out-of-date
+        super().async_write_ha_state()
```

Comment:
State is already written in `super()._handle_coordinator_update()` so move that call to the button and remove the `super().async_write_ha_state()`

Reply:
I always hated doing it, when the `super()` is at the bottom!

In this case, why using `super()` instead of `self` at the end was a type.